frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinTablePicker/JoinTablePicker.tsx

```diff
@@ -21,84 +24,88 @@ import {
 
 interface JoinTablePickerProps {
   query: Lib.Query;
+  stageIndex: number;
   table: Lib.Joinable | undefined;
   tableName: string | undefined;
   color: string;
   isReadOnly: boolean;
-  isModelDataSource: boolean;
   columnPicker: ReactNode;
   onChange?: (table: Lib.Joinable) => void;
 }
 
 export function JoinTablePicker({
   query,
-  table,
+  stageIndex,
+  table: joinable,
   tableName,
   color,
   isReadOnly,
-  isModelDataSource,
   columnPicker,
   onChange,
 }: JoinTablePickerProps) {
-  const metadata = useSelector(getMetadata);
   const dispatch = useDispatch();
+  const onChangeRef = useLatest(onChange);
+  const queryRef = useLatest(query);
 
-  const databaseId = useMemo(() => {
-    return Lib.databaseID(query);
-  }, [query]);
-
-  const databases = useMemo(() => {
-    const database = metadata.database(databaseId);
-    return [database, metadata.savedQuestionsDatabase()].filter(Boolean);
-  }, [databaseId, metadata]);
-
-  const pickerInfo = useMemo(() => {
-    return table ? Lib.pickerInfo(query, table) : null;
-  }, [query, table]);
+  const [isDataPickerOpen, setIsDataPickerOpen] = useState(!joinable);
+  const databaseId = useMemo(() => Lib.databaseID(query), [query]);
 
-  const tableId = pickerInfo?.tableId ?? pickerInfo?.cardId;
-  const tableFilter = (table: Table) => !tableId || table.db_id === databaseId;
   const isDisabled = isReadOnly;
 
   const handleTableChange = async (tableId: TableId) => {
+    // we need to populate query metadata with selected table
     await dispatch(Tables.actions.fetchMetadata({ id: tableId }));
-    onChange?.(Lib.tableOrCardMetadata(query, tableId));
+
+    if (typeof tableId === "string") {
+      await dispatch(
+        Questions.actions.fetch({
+          id: getQuestionIdFromVirtualTableId(tableId),
+        }),
+      );
+    }
```

Comment:
> Hacky way to include changes from metrics-v2 branch without changing many things

See [Slack](https://metaboat.slack.com/archives/C0645JP1W81/p1715268676651349?thread_ts=1715258111.208389&cid=C0645JP1W81).

Without it `Lib.displayInfo` and `Lib.pickerInfo` won't properly recognize the joined thing as a model, and `isModel` will not be `true`.